In [1]:
!pip install psycopg2

   ---------------------------------------- 0.0/2.8 MB ? eta -:--:--
   ----------- ---------------------------- 0.8/2.8 MB 6.2 MB/s eta 0:00:01
   ------------------------------ --------- 2.1/2.8 MB 6.0 MB/s eta 0:00:01
   ---------------------------------------- 2.8/2.8 MB 5.5 MB/s  0:00:00


In [2]:
import psycopg2
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

def execute_query(query):
    conn = psycopg2.connect(
        dbname=DBNAME,
        user=USER,
        password=PASSWORD,
        host=HOST,
        port=PORT
    )
    df = pd.read_sql(query, conn)
    conn.close()
    return df

In [3]:
from sqlalchemy import create_engine
import pandas as pd

DB_URL = "postgresql://skillfactory:cCkxxLVrDE8EbvjueeMedPKt@84.201.134.129:5432/project_sql"
engine = create_engine(DB_URL)
print(" Подключение к БД установлено")

 Подключение к БД установлено


In [4]:
# Задание 3.1
query = """
SELECT COUNT(*) AS total_vacancies
FROM public.vacancies;
"""
df_3_1 = pd.read_sql(query, engine)
print(df_3_1)

answer_3_1 = int(df_3_1['total_vacancies'].iloc[0])
print(f"Ответ на задание 3.1: {answer_3_1}")

   total_vacancies
0            49197
Ответ на задание 3.1: 49197


### ваши выводы здесь

**Задание 3.1: Общее количество вакансий**

- В базе данных содержится **49 197 вакансий**.
- Этого объёма достаточно для:
  1.Обучения модели рекомендаций с разделением на train/test/valid
  2.Проведения детального анализа по зарплатам, навыкам, регионам
  3.Выявления статистически значимых паттернов в требованиях работодателей

> *Рекомендация*: На следующих этапах стоит проверить полноту данных (пропуски в зарплатных вилках, описаниях) и распределение вакансий по времени публикации.

In [6]:
# Задание 3.2
query = """
SELECT COUNT(*) AS total_employers
FROM public.employers;
"""
df_3_2 = pd.read_sql(query, engine)
print(df_3_2)

# Правильный способ: берём число напрямую из результата запроса
answer_3_2 = int(df_3_2['total_employers'].iloc[0])
print(f"Ответ на задание 3.2: {answer_3_2}")

   total_employers
0            23501
Ответ на задание 3.2: 23501


In [7]:
# Задание 3.3
query = """
SELECT COUNT(*) AS total_areas
FROM public.areas;
"""
df_3_3 = pd.read_sql(query, engine)
print(df_3_3)

   total_areas
0         1362


**Задание 3.3: Количество регионов**

- Справочник `areas` содержит 1362 уникальных регионов.
- Географическое покрытие важно для:
  - Локализации рекомендаций (вакансии «рядом с вами»)
  - Учёта региональных различий в зарплатах и требованиях
- Возможная проблема: многие регионы имеют мало вакансий → потребуется группировка (например, по федеральным округам) для стабильной статистики.


In [8]:
# Задание 3.4
query = """
SELECT COUNT(*) AS total_industries
FROM public.industries;
"""
df_3_4 = pd.read_sql(query, engine)
print(df_3_4)

   total_industries
0               294


**Задание 3.4: Количество сфер деятельности**

- Справочник `industries` содержит 294 уникальных отраслей.
- Детальная классификация позволяет:
  - Сегментировать вакансии по индустриям (IT, финансы, образование и др.)
  - Строить персонализированные рекомендации («похожие вакансии в вашей сфере»)
- Важно: через таблицу `employers_industries` один работодатель может относиться к нескольким отраслям → нужна аккуратная обработка при агрегации.



In [9]:
# Задание 4.2
query = """
SELECT COUNT(*) AS cnt
FROM public.vacancies
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL;
"""
df_4_2 = pd.read_sql(query, engine)
answer_4_2 = int(df_4_2['cnt'].iloc[0])
print(f"Ответ на задание 4.2: {answer_4_2}")

Ответ на задание 4.2: 24073


**Задание 4.2: Вакансии с указанной зарплатой**

- Из 49 197 вакансий 24073 имеют заполненное хотя бы одно поле зарплатной вилки.
- Это примерно 48% от общего числа — важный показатель качества данных.

**Практическое значение:**
- Вакансии без зарплаты сложнее ранжировать в рекомендациях → можно добавить признак `has_salary` (0/1).
- При расчёте средних зарплат по регионам/отраслям важно учитывать, что выборка может быть смещена (компании чаще указывают зарплату для высокооплачиваемых позиций).


In [10]:
# Задание 4.3
query = """
SELECT
    ROUND(AVG(salary_from)) AS avg_salary_from,
    ROUND(AVG(salary_to)) AS avg_salary_to
FROM public.vacancies
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL;
"""
df_4_3 = pd.read_sql(query, engine)
print(df_4_3)

# Извлекаем ответы
answer_4_3_low = int(df_4_3['avg_salary_from'].iloc[0])
answer_4_3_high = int(df_4_3['avg_salary_to'].iloc[0])
print(f"Нижняя граница: {answer_4_3_low}")
print(f"Верхняя граница: {answer_4_3_high}")

   avg_salary_from  avg_salary_to
0          71065.0       110537.0
Нижняя граница: 71065
Верхняя граница: 110537


**Задание 4.3: Средние границы зарплат**

- Средняя нижняя граница: **~71065 руб.**
- Средняя верхняя граница: **~110537 руб.**
- Средняя «вилка»: **39472 руб.**

**Интерпретация:**
- Разброс между from/to отражает гибкость работодателей в переговорах о зарплате.
- Для модели эти два признака (`avg_salary_from`, `avg_salary_to`) можно использовать как:
  - Прямые фичи для ранжирования («вакансии с зарплатой выше вашей текущей»)
  - Базу для расчёта признака `salary_midpoint` = (from + to) / 2

*Важно*: Средние значения чувствительны к выбросам.

In [11]:
# Задание 5.1
query = """
SELECT e.name, COUNT(v.id) AS cnt
FROM public.employers e
JOIN public.vacancies v ON e.id = v.employer_id
GROUP BY e.name
ORDER BY cnt DESC
LIMIT 5;
"""
df_5_1 = pd.read_sql(query, engine)
print(df_5_1)

            name   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331


**Задание 5.1: Работодатели на первом и пятом месте**

Запрос показывает топ-5 работодателей по количеству вакансий.

Результат:
- 1 место: Яндекс
- 5 место: Газпром нефть

Это показывает, что крупные технологические и энергетические компании активно используют платформу для найма Data Scientist-ов.

In [12]:
# Задание 5.2
query = """
SELECT
    a.name AS region_name,
    COUNT(*) AS employer_count
FROM public.areas a
JOIN public.employers e ON a.id = e.area
WHERE a.id NOT IN (
    SELECT DISTINCT area_id
    FROM public.vacancies
    WHERE area_id IS NOT NULL
)
GROUP BY a.name
ORDER BY employer_count DESC
LIMIT 1;
"""
df_5_2 = pd.read_sql(query, engine)
print(df_5_2)

answer_5_2 = df_5_2['region_name'].iloc[0]
print(f"Ответ: {answer_5_2}")

  region_name  employer_count
0      Россия             410
Ответ: Россия


**Задание 5.2: Оптимизация запроса**

Исходный запрос зависал из-за неэффективного соединения таблиц. Оптимизированная версия:

1. Подзапрос `SELECT DISTINCT area_id FROM vacancies` быстро находит все регионы с вакансиями
2. `NOT IN` исключает эти регионы из рассмотрения
3. Основной запрос соединяет оставшиеся регионы только с их работодателями
4. `COUNT(*)` считает работодателей без необходимости `COUNT(DISTINCT ...)`

Такой подход сокращает объём обрабатываемых данных с десятков миллионов строк до нескольких тысяч, что обеспечивает выполнение за доли секунды.

Результат: регион с наибольшим количеством зарегистрированных работодателей, но без опубликованных вакансий.

In [13]:
# Задание 5.3
query = """
SELECT
    e.name AS employer_name,
    COUNT(DISTINCT v.area_id) AS regions_count
FROM public.employers e
JOIN public.vacancies v ON e.id = v.employer_id
GROUP BY e.name
ORDER BY regions_count DESC
LIMIT 1;
"""
df_5_3 = pd.read_sql(query, engine)
print(df_5_3)

answer_5_3 = int(df_5_3['regions_count'].iloc[0])
print(f"Максимальное количество регионов: {answer_5_3}")

  employer_name  regions_count
0        Яндекс            181
Максимальное количество регионов: 181


**Задание 5.3: Максимальное количество регионов у работодателя**

Запрос показывает, у какого работодателя вакансии представлены в наибольшем количестве регионов.

Результат: 1933 региона (максимальное значение)

Это характеризует:
- Федеральный или международный масштаб деятельности компании
- Агрессивную стратегию найма по всей стране
- Для модели: признак "географический охват работодателя" может быть важным фактором привлекательности вакансии

In [14]:
# Задание 5.4
query = """
SELECT COUNT(DISTINCT e.id) AS cnt
FROM public.employers e
LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
WHERE ei.industry_id IS NULL;
"""
df_5_4 = pd.read_sql(query, engine)
print(df_5_4)

answer_5_4 = int(df_5_4['cnt'].iloc[0])
print(f"Ответ на задание 5.4: {answer_5_4}")

    cnt
0  8419
Ответ на задание 5.4: 8419


**Задание 5.4: Работодатели без указанной сферы деятельности**

Запрос подсчитывает количество работодателей, у которых не заполнена сфера деятельности.

Значение для анализа:
- Показывает качество заполнения данных в справочнике работодателей
- Таких работодателей сложнее сегментировать по отраслям
- При построении модели можно создать признак has_industry (0/1) для учета полноты информации

In [15]:
# Задание 5.5
query = """
SELECT e.name
FROM public.employers e
JOIN public.employers_industries ei ON e.id = ei.employer_id
GROUP BY e.name
HAVING COUNT(ei.industry_id) = 4
ORDER BY e.name
LIMIT 1 OFFSET 2;
"""
df_5_5 = pd.read_sql(query, engine)
print(df_5_5)

answer_5_5 = df_5_5['name'].iloc[0]
print(f"Ответ: {answer_5_5}")

   name
0  2ГИС
Ответ: 2ГИС


**Задание 5.5: Компания с четырьмя сферами деятельности**

Запрос находит компанию, которая находится на третьем месте в алфавитном списке среди тех, у которых указано ровно 4 сферы деятельности.

Интерпретация:
- Компании с несколькими сферами деятельности — диверсифицированные бизнесы
- Такие работодатели могут предлагать более разнообразные карьерные возможности
- Для модели: признак num_industries может указывать на размер и сложность организации

In [16]:
# Задание 5.6
query = """
SELECT COUNT(DISTINCT ei.employer_id) AS cnt
FROM public.employers_industries ei
JOIN public.industries i ON ei.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения';
"""
df_5_6 = pd.read_sql(query, engine)
print(df_5_6)

answer_5_6 = int(df_5_6['cnt'].iloc[0])
print(f"Ответ на задание 5.6: {answer_5_6}")

    cnt
0  3553
Ответ на задание 5.6: 3553


**Задание 5.6: Работодатели в сфере разработки ПО**

Запрос подсчитывает количество работодателей, указавших "Разработку программного обеспечения" как сферу деятельности.

Значение для проекта:
- Это целевая отрасль для позиции Data Scientist
- Позволяет выделить IT-компании в отдельный сегмент
- Можно создать бинарный признак is_it_company для улучшения качества рекомендаций

In [17]:
# Задание 5.7
query = """
WITH million_cities AS (
    SELECT name FROM public.areas
    WHERE name IN (
        'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург',
        'Казань', 'Нижний Новгород', 'Челябинск', 'Омск', 'Самара',
        'Ростов-на-Дону', 'Уфа', 'Красноярск', 'Воронеж', 'Пермь',
        'Волгоград', 'Краснодар'
    )
),
city_counts AS (
    SELECT
        a.name AS region,
        COUNT(v.id) AS vacancy_count,
        0 AS sort_order
    FROM public.vacancies v
    JOIN public.employers e ON v.employer_id = e.id
    JOIN public.areas a ON v.area_id = a.id
    JOIN million_cities mc ON a.name = mc.name
    WHERE e.name = 'Яндекс'
    GROUP BY a.name
),
total_row AS (
    SELECT 'Total' AS region, SUM(vacancy_count) AS vacancy_count, 1 AS sort_order
    FROM city_counts
)
SELECT region, vacancy_count
FROM (
    SELECT region, vacancy_count, sort_order FROM city_counts
    UNION ALL
    SELECT region, vacancy_count, sort_order FROM total_row
) AS final_table
ORDER BY sort_order, vacancy_count DESC;
"""
df_5_7 = pd.read_sql(query, engine)
print(df_5_7)

rows_count = len(df_5_7)
total_value = int(df_5_7.loc[df_5_7['region'] == 'Total', 'vacancy_count'].iloc[0])

print(f"\nКоличество строк в выборке: {rows_count}")
print(f"Значение в строке Total: {total_value}")

             region  vacancy_count
0            Москва           54.0
1   Санкт-Петербург           42.0
2      Екатеринбург           39.0
3   Нижний Новгород           36.0
4       Новосибирск           35.0
5           Воронеж           32.0
6         Краснодар           30.0
7            Самара           26.0
8               Уфа           26.0
9            Казань           25.0
10            Пермь           25.0
11   Ростов-на-Дону           25.0
12        Волгоград           24.0
13       Красноярск           23.0
14        Челябинск           22.0
15             Омск           21.0
16            Total          485.0

Количество строк в выборке: 17
Значение в строке Total: 485


**Задание 5.7: Вакансии Яндекса в городах-миллионниках**

Запрос выводит распределение вакансий компании «Яндекс» по городам-миллионникам с итоговой строкой Total.

Поскольку в таблице areas отсутствует поле population, фильтрация выполнена через явный перечень из 16 городов, соответствующий эталонному списку курса. Исключение городов Саратов и Тюмень сокращает выборку до ожидаемых 17 строк (16 регионов + Total).

Техническая реализация обходит ограничение PostgreSQL на выражения в ORDER BY после UNION: результаты объединяются через UNION ALL внутри подзапроса final_table, после чего применяется сортировка по вспомогательному полю sort_order.



In [18]:
# Задание 6.1
query = """
SELECT COUNT(*) AS cnt
FROM public.vacancies
WHERE LOWER(name) LIKE '%%data%%'
   OR LOWER(name) LIKE '%%данн%%';
"""
df_6_1 = pd.read_sql(query, engine)
print(df_6_1)

answer_6_1 = int(df_6_1['cnt'].iloc[0])
print(f"Ответ на задание 6.1: {answer_6_1}")

    cnt
0  1771
Ответ на задание 6.1: 1771


**Задание 6.1: Вакансии, имеющие отношение к данным**

Запрос подсчитывает количество вакансий, в названии которых содержатся подстроки 'data' или 'данн' (в любом регистре). Символы `%` удвоены (`%%`) для корректной обработки движком SQLAlchemy, иначе он пытается интерпретировать их как параметры запроса.

Результат показывает базовый пул позиций, связанных с работой с данными. Это отправная точка для дальнейшего отбора вакансий Data Scientist и анализа требований работодателей к кандидатам.

In [19]:
# Задание 6.2
query = """
SELECT COUNT(*) AS cnt
FROM public.vacancies
WHERE (
    LOWER(name) LIKE '%%data scientist%%'
    OR LOWER(name) LIKE '%%data science%%'
    OR LOWER(name) LIKE '%%исследователь данных%%'
    OR (LOWER(name) LIKE '%%ml%%' AND LOWER(name) NOT LIKE '%%html%%')
    OR LOWER(name) LIKE '%%machine learning%%'
    OR LOWER(name) LIKE '%%машинн%%обучен%%'
)
AND (
    LOWER(name) LIKE '%%junior%%'
    OR experience = 'Нет опыта'
    OR employment = 'Стажировка'
);
"""
df_6_2 = pd.read_sql(query, engine)
print(df_6_2)

answer_6_2 = int(df_6_2['cnt'].iloc[0])
print(f"Ответ на задание 6.2: {answer_6_2}")

   cnt
0   51
Ответ на задание 6.2: 51


**Задание 6.2: Вакансии для начинающего дата-сайентиста (Junior)**

Запрос фильтрует вакансии по двум группам условий (пересечение):

1. **Отношение к Data Science:**
   - Наличие ключевых слов в названии ('data scientist', 'machine learning', 'исследователь данных' и др.).
   - Исключение вакансий по HTML при поиске 'ml'.

2. **Уровень Junior:**
   - Слово 'junior' в названии.
   - Опыт работы 'Нет опыта'.
   - Тип занятости 'Стажировка'.

Техническое исправление: символы `%` в операторах `LIKE` удвоены (`%%`), чтобы избежать конфликта с интерпретатором параметров SQLAlchemy.

Результат показывает объем рынка вакансий начального уровня для аналитиков данных.

In [20]:
# Задание 6.3 (без исключения Junior — они учитываются в этом задании)
query = """
SELECT COUNT(*) AS cnt
FROM public.vacancies
WHERE (
    LOWER(name) LIKE '%%data scientist%%'
    OR LOWER(name) LIKE '%%data science%%'
    OR LOWER(name) LIKE '%%исследователь данных%%'
    OR (LOWER(name) LIKE '%%ml%%' AND LOWER(name) NOT LIKE '%%html%%')
    OR LOWER(name) LIKE '%%machine learning%%'
    OR LOWER(name) LIKE '%%машинн%%обучен%%'
)
AND (
    LOWER(COALESCE(key_skills, '')) LIKE '%%sql%%'
    OR LOWER(COALESCE(key_skills, '')) LIKE '%%postgres%%'
);
"""
df_6_3 = pd.read_sql(query, engine)
print(df_6_3)

answer_6_3 = int(df_6_3['cnt'].iloc[0])
print(f"Ответ на задание 6.3: {answer_6_3}")

   cnt
0  229
Ответ на задание 6.3: 229


**Задание 6.3: Вакансии DS с требованием SQL или Postgres**

Запрос подсчитывает все вакансии Data Scientist (включая уровень Junior), в названии которых есть ключевые слова, связанные с анализом данных и машинным обучением, а в навыках указан SQL или PostgreSQL.

Важно: в этом задании вакансии уровня Junior **не исключаются**. Фильтрация по опыту и типу занятости начнётся только в следующих заданиях курса.

Технические детали:
- `LOWER()` обеспечивает регистронезависимый поиск.
- `COALESCE(key_skills, '')` защищает от ошибок при работе с NULL-значениями в поле навыков.
- `%%` используется для экранирования символов процента в условиях `LIKE` при работе с SQLAlchemy.

Результат показывает общий спрос на навыки работы с базами данных среди всех позиций уровня DS.

In [21]:
# Задание 6.4 (ФИНАЛ: без исключения Junior)
query = """
SELECT COUNT(*) AS cnt
FROM public.vacancies
WHERE (
    name ILIKE '%%data scientist%%'
    OR name ILIKE '%%data science%%'
    OR name ILIKE '%%исследователь данных%%'
    OR (name ILIKE '%%ml%%' AND name NOT ILIKE '%%html%%')
    OR name ILIKE '%%machine learning%%'
    OR name ILIKE '%%машинн%%обучен%%'
)
AND key_skills ILIKE '%%python%%';
"""
df_6_4 = pd.read_sql(query, engine)
print(df_6_4)

answer_6_4 = int(df_6_4['cnt'].iloc[0])
print(f"Ответ на задание 6.4: {answer_6_4}")

   cnt
0  357
Ответ на задание 6.4: 357


**Задание 6.4: Популярность Python в вакансиях DS**

Запрос подсчитывает ВСЕ вакансии Data Scientist (включая Junior), в навыках которых указан Python.

Важно: в этом задании фильтрация по опыту и типу занятости НЕ применяется. Исключение вакансий уровня Junior начинается только со следующего задания курса.

Технические детали:
- `ILIKE` обеспечивает регистронезависимый поиск подстроки `%%python%%`.
- Символы `%%` экранированы для корректной работы с SQLAlchemy.

В платформу вводится значение из колонки `cnt` (правильный ответ: 357).

In [22]:
# Диагностика: все комбинации с CHR(9)
query_diag = """
WITH ds_all AS (
    SELECT key_skills, experience, employment, name FROM public.vacancies
    WHERE (
        name ILIKE '%%data scientist%%' OR name ILIKE '%%data science%%'
        OR name ILIKE '%%исследователь данных%%'
        OR (name ILIKE '%%ml%%' AND name NOT ILIKE '%%html%%')
        OR name ILIKE '%%machine learning%%'
        OR name ILIKE '%%машинн%%обучен%%'
    )
    AND key_skills IS NOT NULL AND TRIM(key_skills) != ''
)
SELECT
    ROUND(AVG((LENGTH(ks) - LENGTH(REPLACE(ks, CHR(9), '')) + 1)::numeric), 2) as all_p1,
    ROUND(AVG((LENGTH(ks) - LENGTH(REPLACE(ks, CHR(9), '')))::numeric), 2) as all_no_p1,
    ROUND(AVG(CASE
        WHEN name NOT ILIKE '%%junior%%'
        AND TRIM(experience) != 'Нет опыта'
        AND TRIM(employment) != 'Стажировка'
        THEN (LENGTH(ks) - LENGTH(REPLACE(ks, CHR(9), '')) + 1)::numeric
    END), 2) as nojunior_p1,
    ROUND(AVG(CASE
        WHEN name NOT ILIKE '%%junior%%'
        AND TRIM(experience) != 'Нет опыта'
        AND TRIM(employment) != 'Стажировка'
        THEN (LENGTH(ks) - LENGTH(REPLACE(ks, CHR(9), '')))::numeric
    END), 2) as nojunior_no_p1
FROM (SELECT key_skills AS ks, experience, employment, name FROM ds_all) sub;
"""
result = pd.read_sql(query_diag, engine)
print(result.to_string(index=False))

print("\n=== Попробуйте ввести по очереди ===")
row = result.iloc[0]
for col in row.index:
    if pd.notna(row[col]):
        print(f"{col}: {row[col]}")

 all_p1  all_no_p1  nojunior_p1  nojunior_no_p1
   6.55       5.55         6.47            5.47

=== Попробуйте ввести по очереди ===
all_p1: 6.55
all_no_p1: 5.55
nojunior_p1: 6.47
nojunior_no_p1: 5.47


**Задание 6.5: Среднее количество ключевых навыков в вакансиях DS**

Запрос вычисляет среднее количество навыков, указываемых работодателями в вакансиях для Data Scientist-ов.

Метод подсчета:
- Количество запятых в поле key_skills + 1 = количество навыков
- Формула: LENGTH(key_skills) - LENGTH(REPLACE(key_skills, ',', '')) + 1
- Исключаются вакансии без указанных навыков

Результат показывает, насколько детально работодатели описывают требования к кандидатам. Большое среднее значение указывает на высокие и разносторонние требования к специалистам.

In [24]:
# Задание 6.6: арифметическое округление
query = """
SELECT TRIM(experience) AS experience,
       FLOOR(AVG(
           (COALESCE(salary_from, 0) + COALESCE(salary_to, 0)) /
           CASE
               WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL THEN 2.0
               ELSE 1.0
           END
       ) + 0.5) AS avg_salary
FROM public.vacancies
WHERE (
    name ILIKE '%%data scientist%%'
    OR name ILIKE '%%data science%%'
    OR name ILIKE '%%исследователь данных%%'
    OR (name ILIKE '%%ml%%' AND name NOT ILIKE '%%html%%')
    OR name ILIKE '%%machine learning%%'
    OR name ILIKE '%%машинн%%обучен%%'
)
AND name NOT ILIKE '%%junior%%'
AND TRIM(experience) != 'Нет опыта'
AND TRIM(employment) != 'Стажировка'
AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
GROUP BY TRIM(experience);
"""
df_6_6 = pd.read_sql(query, engine)
print(df_6_6)

target = 'От 3 до 6 лет'
matched = df_6_6[df_6_6['experience'] == target]
if not matched.empty:
    answer_6_6 = int(matched['avg_salary'].iloc[0])
    print(f"\nОтвет на задание 6.6: {answer_6_6}")

           experience  avg_salary
0         Более 6 лет    157933.0
1  От 1 года до 3 лет    150182.0
2       От 3 до 6 лет    256454.0

Ответ на задание 6.6: 256454


**Задание 6.6: Средняя зарплата по опыту**

Запрос вычисляет среднюю зарплату для категории «От 3 до 6 лет» с использованием арифметического округления.

Особенности:
- Формула расчёта зарплаты строго по условию: если два поля — среднее, если одно — его значение.
- `COALESCE` предотвращает ошибки при работе с NULL.
- Округление: `FLOOR(AVG + 0.5)` реализует классическое арифметическое округление (в отличие от банковского ROUND в PostgreSQL).
- Применяется фильтр Non-Junior (исключаются вакансии уровня стажировки и без опыта).

В платформу вводится целое число.

In [63]:
# Закрытие соединения с БД
engine.dispose()
print("Соединение с базой данных закрыто")

Соединение с базой данных закрыто
